Create CloudFormation template SSM Run Commands

jkataja opened the pull request:
Upload all of the SSM Documents using CloudFormation template created from the SSM Run Document files.

 - Reads YAML files from input path specified in first argument.
 - Command expects the input path contains SSM Run Command documents.
 - Outputs CloudFormation YAML template with all the SSM Run Commands.

Example usage:

```
$ cd chaos-ssm-documents/
$ run-command/create-cfn.sh run-command/ | tee cfn-chaos-ssm.yml
$ aws cloudformation create-stack --stack-name ChaosSsm --template-body file://cfn-chaos-ssm.yml
```

Benefit over `upload-document.sh` is being able to clean and update the run documents more easily, and to deploy them using Stack Sets.

jkataja committed:
Create CloudFormation template SSM Run Commands

jkataja committed:
improved comments

adhorn commented:
Thanks for the contribution - However, the current yml generated file isn't valid YAML but generates an extra `:` in the `Resource:`

```
AWSTemplateFormatVersion: 2010-09-09

Description: |
  Chaos Injection for AWS resources using Amazon SSM Run Command and Automation

Resources:

  # blackhole-dns-stress.yml
  :
    Type: AWS::SSM::Document
    Properties:
      DocumentType: Command
```
The `:` is happening on a clean checkout of the branch. 

Using a Mac FYI.

Other than that, loving the idea!

jkataja commented:
Thanks for the quick feedback. Liked your talk on re:Invent  about this, because it was easy and practical approach to chaos engineering. I will take a look later today - sounds like the Mac sed might not accept all the args. I developed this on Fedora Linux.

adhorn commented:
Thank you! Yea, that's what I suspected. `sed`on Mac is peculiar. Let me know if you figure it out, I can take some time later today to look at it too. 
Thanks

jkataja committed:
perl instead of sed for resource name CamelCase

 - sed was not compatible with macOS sed

jkataja commented:
I have removed sed and replaced with with call to perl to convert resource name to CamelCase

adhorn commented:
Thanks - works great now. 
Do you typically you typically keep the generated file in the repo? I tend not to in which case it would be good to add it to `.gitignore`- What do you think?

adhorn commented:
Another comment - sorry - when updating a template and running `update-stack`the name of the resource created changes, which makes it hard to integrate the SSM document via `name`. 
Could the script add the name of the resource? 
```
Type: AWS::SSM::Document
Properties: 
  Content: Json
  DocumentType: String
  Name: String
  Tags: 
    - Tag
```
So that `update-stack`creates a new version of the document instead of deleting the old one and replacing it.

Comments?

jkataja commented:
> Thanks - works great now.

No problem

> Do you typically you typically keep the generated file in the repo? I tend not to in which case it would be good to add it to `.gitignore`- What do you think?

The generated file name is not fixed in the script, however top level `.gitignore` could filter all `yaml` and `yml` files unless specifically added

jkataja commented:
> Another comment - sorry - when updating a template and running `update-stack`the name of the resource created changes, which makes it hard to integrate the SSM document via `name`.

Once a CloudFormation stack is created, the resource names should not change anymore when updating the stack. Or there could be a bug on the SSM resources in CF?

> Could the script add the name of the resource?

I usually let CloudFormation generate the resource names as much as possible, because then there are no chances of naming collisions. The downside is that the names are now longer and not always the same. Which would you prefer?

> So that `update-stack`creates a new version of the document instead of deleting the old one and replacing it.

I have to test the update first., just assumed that update would work without replacing the resources. If updating the stack replaces all the resources then might be a bug in CF.

Also, maybe the documented stack name could be `Chaos` to keep the name shorter.

jkataja commented:
- [x] .gitignore to all yaml and yml files
- [x] test update stack works without replacing the documents

adhorn commented:
> I usually let CloudFormation generate the resource names as much as possible, because then there are no chances of naming collisions. The downside is that the names are now longer and not always the same. Which would you prefer?
> 
> > So that `update-stack`creates a new version of the document instead of deleting the old one and replacing it.

I really don't mind if it is auto-generated as long as it the generated name contains the name so that it is obvious when debugging what script was run, and as long as the name never changes after an update. 

How did you envision the update to happen? re-run the `run-command/create-cfn.sh run-command/ | tee cfn-chaos-ssm.yml`?

adhorn commented:
> > Another comment - sorry - when updating a template and running `update-stack`the name of the resource created 
> I have to test the update first., just assumed that update would work without replacing the resources. If updating the stack replaces all the resources then might be a bug in CF.

`AWS CloudFormation recreates the resource during an update, which also generates a new physical ID. AWS CloudFormation creates the replacement resource first, changes references from other dependent resources to point to the replacement resource, and then deletes the old resource. For example, if you update the Engine property of an AWS::RDS::DBInstance resource type, AWS CloudFormation creates a new resource and replaces the current DB instance resource with the new one. `

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/using-cfn-updating-stacks-update-behaviors.html#update-replacement

That's why I suggested adding the name - so the reference to the template is constant.

jkataja committed:
add YAML files to gitignore

 - avoid generated CloudFormation YAML files to be committed

jkataja commented:
> How did you envision the update to happen? re-run the `run-command/create-cfn.sh run-command/ | tee cfn-chaos-ssm.yml`?

Re-running the command and then using `aws cloudformation update-stack`. After modifying one of the documents, recreating the stack and then running `update-stack`, I now see that indeed it creates a new resource and then deletes the old one. I did not expect behavior this but indeed it works like you said. As the end result the resource name is changed.
 
With named SSM documents, I get an an error with the following status when trying to update the stack:

> CloudFormation cannot update a stack when a custom-named resource requires replacing. Rename Chaos-InstallDependencies and update the stack again.

Update process is remove old and re-deploy the new stack. Fortunately it deploys very quickly. I added `Name `field to documents so the names stay the same.

jkataja committed:
add comment on updating CloudFormation stack

 - updating SSM Document resources requires remove and create
 - resource is created with a new name if name is undefined,
   with fixeed naming stack update will fail to apply update

jkataja committed:
add resource name to SSM Document definition

jkataja committed:
add comment on update-stack not working

jkataja committed:
out GitHub repo and head commit in generated file

adhorn commented:
Perfect - thanks @jkataja  for the contribution!
